event-store-plans.md

```diff
@@ -0,0 +1,174 @@
+work with randomness to kind of distribute the load per tenant
+
+By no means will this be a true equivalent of the projection load balancing work planned for "CritterStackPro"
+
+## Dynamic Database per Tenant Multi-Tenancy model
+
+A JasperFx client is sponsoring this work, so it's absolutely in scope. This time there'll need to be a "master" database that has a table
+to track tenants and tenant databases. The async daemon should be able to discover new tenant databases at runtime and try to spin up a new daemon
+for each tenant as a way to de facto discover and activate new tenants.
+
+
+## "First Class Subscription" model
+
+A new model for subscriptions through the async daemon as described by Oskar here: [https://event-driven.io/en/integrating_Marten/](https://event-driven.io/en/integrating_Marten/)
+
+I think this deserves a slightly different abstraction like:
+
+```csharp
+public interface IProjection
+{
+    Task SubscribeAsync(IDocumentOperations operations, IReadOnlyList<IEvent> events,
+        CancellationToken cancellation);
+}
+```
+
+Behind the scenes it's really just just being driven by the daemon running a single page of events at a time, and recording
+the process as it does with projections today.
+
+Now, other issues to consider:
+
+* Is this done as a "hot" subscription that starts wherever the event high water mark is already when a new one is added?
+* Do we support "cold" subscriptions that would cause a new subscription to start from the the beginning at event #1?
+* How is `ISubscription` handled in rebuilds? Do we have specific replays? How do we help users keep from doing this accidentally when they're rebuilding projections?
+
+
+
+
+## FetchForReading alternative
+
+Same functionality as `FetchForWriting()`, but leave out a little bit of the versioning functionality as this would be 
+read-only. Encourage users to avoid using `LoadAsync()` or `AggregateStreamAsync()` by hand so that they can more easily
+switch between projection lifecycles. This will be almost absolutely necessary when we go to the "zero downtime" projection
+rebuilds
+
+## Async strategy for FetchForWriting()
+
+What I've been calling the "catch up" mode. Essentially [this issue](https://github.com/JasperFx/marten/issues/2846). If a per stream aggregation projection
+is running in async mode, calling `FetchForWriting()` would fetch the current version of the projected document and
+all the events *after* the saved version of the aggregate document in a single database call and apply them on top
+of the saved aggregate
+
+I think this is going to require some additional options for "versioned documents". Instead of the `Guid`-based strategy we just today for optimistic versioning, we'd 
+need an `int` based model where the stream version is also embedded into the document storage for the projected document. 
+
```

Comment:
(Tagging a random line near the bottom so that discussion threads are possible)

Additional possible idea: **Remove the need for multi-stream projections to use the same ID type as is configured for the event store.**

Reasoning here is that if you use guids for stream keys and you have to combine streams in some way for a multi-stream projection, there is a good chance that the resulting grain of the projection is not going to be a guid.
The only reason I am not using guid stream keys for events is so that multi-stream projections can support string ids. The limit feels arbitrary and I suspect is not their as intentional design.